# Project: Loan Default Risk Analysis (Lending Club + CA Unemployment)
# Notebook: 01_data_audit_and_understanding.ipynb
# Purpose: Raw data loading and audit
# Author: Ghayasudin Ghayas
# Date: 2026-01-1

In [1]:
from google.colab import drive
drive.mount('/content/drive')

PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


#Step 1: Set the project root

In [7]:
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"

RAW_PATH = PROJECT_ROOT + "data/raw/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"
OUTPUTS_PATH = PROJECT_ROOT + "outputs/"

import os
for path in [RAW_PATH, PROCESSED_PATH, FILTERED_PATH, OUTPUTS_PATH]:
    print(path, "→", os.path.exists(path))


/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/raw/ → True
/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/processed/ → True
/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/ → True
/content/drive/MyDrive/LOAN_DEFAULT_RISK/outputs/ → True


# Step 4: Import libraries

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", 200)
pd.set_option("display.max_rows", 200)


# Step 5: Load your datasets

In [9]:
accepted_path = RAW_PATH + "accepted_2007_to_2018Q4.csv"
#rejected_path = RAW_PATH + "rejected_2007_to_2018Q4.csv"
#unemp_path    = RAW_PATH + "ca_unemployment_seasonally_adjusted.xlsx"

accepted_df = pd.read_csv(accepted_path, low_memory=False)
#rejected_df = pd.read_csv(rejected_path, low_memory=False)
#unemp_df    = pd.read_excel(unemp_path)

print("Accepted:", accepted_df.shape)
#print("Rejected:", rejected_df.shape)
#print("Unemployment:", unemp_df.shape)


Accepted: (2260701, 151)


In [10]:
accepted_df.head()
accepted_df.info()
# accepted_df.describe(include="all").T


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260701 entries, 0 to 2260700
Columns: 151 entries, id to settlement_term
dtypes: float64(113), object(38)
memory usage: 2.5+ GB


In [ ]:
accepted_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260701 entries, 0 to 2260700
Columns: 151 entries, id to settlement_term
dtypes: float64(113), object(38)
memory usage: 2.5+ GB


In [11]:
# Filter California only
accepted_ca_df = accepted_df[accepted_df['addr_state'] == 'CA']

print("California dataset shape:", accepted_ca_df.shape)

accepted_ca_df.head()



California dataset shape: (314533, 151)


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
13,66624733,NaN,18000.0,18000.0,18000.0,60 months,19.48,471.70,E,E2,Software Manager,7 years,RENT,150000.0,Not Verified,Dec-2015,Charged Off,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,debt_consolidation,Debt consolidation,916xx,CA,9.39,0.0,Jun-2005,665.0,669.0,1.0,NaN,75.0,18.0,1.0,14052.0,40.7,27.0,w,0.00,0.00,9452.740000,9452.74,3481.86,4351.98,0.0,1618.9,291.402,May-2017,471.70,NaN,Nov-2017,584.0,580.0,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,60.0,36247.0,2.0,2.0,1.0,3.0,2.0,22195.0,71.0,2.0,4.0,1271.0,55.0,34500.0,2.0,1.0,2.0,7.0,2014.0,5128.0,51.3,0.0,0.0,115.0,112.0,1.0,1.0,2.0,9.0,NaN,2.0,NaN,0.0,6.0,13.0,7.0,8.0,6.0,16.0,19.0,13.0,18.0,0.0,0.0,0.0,3.0,100.0,14.3,1.0,0.0,65819.0,36247.0,10300.0,31319.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
42,68615044,NaN,16000.0,16000.0,16000.0,60 months,13.99,372.21,C,C4,Registered Nurse,9 years,MORTGAGE,142000.0,Verified,Dec-2015,Current,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,debt_consolidation,Debt consolidation,945xx,CA,17.74,2.0,Oct-2004,680.0,684.0,2.0,8.0,39.0,15.0,2.0,2273.0,39.9,19.0,w,7186.06,7186.06,14112.890000,14112.89,8813.94,5298.95,0.0,0.0,0.000,Feb-2019,372.21,Apr-2019,Mar-2019,689.0,685.0,0.0,43.0,1.0,Individual,NaN,NaN,NaN,0.0,0.0,667373.0,0.0,7.0,0.0,6.0,17.0,109893.0,69.0,0.0,1.0,1959.0,66.0,5700.0,5.0,0.0,3.0,7.0,47670.0,2827.0,44.6,0.0,0.0,134.0,98.0,23.0,17.0,1.0,23.0,29.0,4.0,13.0,1.0,2.0,2.0,4.0,5.0,10.0,7.0,8.0,2.0,15.0,0.0,0.0,0.0,0.0,73.7,25.0,0.0,2.0,7

In [13]:
#accepted_df.describe(include="all").T


In [14]:
import pandas as pd

# Basic info table
summary = pd.DataFrame({
    "dtype": accepted_ca_df.dtypes,
    "non_null_count": accepted_ca_df.count(),
    "null_count": accepted_ca_df.isna().sum()
})

summary["null_pct"] = (summary["null_count"] / len(accepted_ca_df)) * 100

summary = summary.sort_values("null_pct", ascending=False)

summary.head(20)   # top 20 most-missing columns


,dtype,non_null_count,null_count,null_pct
member_id,float64,0,314533,100.000000
orig_projected_additional_accrued_interest,float64,731,313802,99.767592
hardship_reason,object,974,313559,99.690335
hardship_payoff_balance_amount,float64,974,313559,99.690335
hardship_last_payment_amount,float64,974,313559,99.690335
payment_plan_start_date,object,974,313559,99.690335
hardship_type,object,974,313559,99.690335
hardship_status,object,974,313559,99.690335
hardship_start_date,object,974,313559,99.690335
deferral_term,float64,974,313559,99.690335


In [15]:
accepted_ca_df.describe(include='all').transpose()


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
id,314533,314533,88985880,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
member_id,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
loan_amnt,314533.0,NaN,NaN,NaN,15287.680784,9307.169748,500.0,8000.0,13000.0,20000.0,40000.0
funded_amnt,314533.0,NaN,NaN,NaN,15280.914721,9305.072342,500.0,8000.0,13000.0,20000.0,40000.0
funded_amnt_inv,314533.0,NaN,NaN,NaN,15260.240877,9309.565055,0.0,8000.0,13000.0,20000.0,40000.0
term,314533,2,36 months,233577,NaN,NaN,NaN,NaN,NaN,NaN,NaN
int_rate,314533.0,NaN,NaN,NaN,12.976528,4.798863,5.31,9.44,12.59,15.61,30.99
installment,314533.0,NaN,NaN,NaN,458.517093,274.74906,16.31,258.1,388.2,615.11,1715.42
grade,314533,7,B,94433,NaN,NaN,NaN,NaN,NaN,NaN,NaN
sub_grade,314533,35,C1,20206,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [16]:
threshold = 50  # percent

cols_to_drop = summary[summary["null_pct"] > threshold].index

accepted_ca_df_reduced = accepted_ca_df.drop(columns=cols_to_drop)

print("Original shape:", accepted_ca_df.shape)
print("Reduced shape :", accepted_ca_df_reduced.shape)
print("Dropped columns:", len(cols_to_drop))


Original shape: (314533, 151)
Reduced shape : (314533, 106)
Dropped columns: 45


In [17]:
text_cols = ['id','member_id','url','desc','title','emp_title','zip_code']
accepted_ca_df_reduced.drop(columns=[c for c in text_cols if c in accepted_ca_df_reduced], inplace=True)


In [18]:
print("Shape after dropping text columns:", accepted_ca_df_reduced.shape)

# Memory usage
print("\nMemory usage after drop:")
accepted_ca_df_reduced.info(memory_usage='deep')


Shape after dropping text columns: (314533, 101)

Memory usage after drop:
<class 'pandas.core.frame.DataFrame'>
Index: 314533 entries, 13 to 2260696
Columns: 101 entries, loan_amnt to debt_settlement_flag
dtypes: float64(82), object(19)
memory usage: 512.2 MB


In [19]:
dropped = set(['id','member_id','url','desc','title','emp_title','zip_code'])
print("Dropped columns:", dropped.intersection(accepted_ca_df.columns))


Dropped columns: {'member_id', 'id', 'title', 'emp_title', 'url', 'zip_code', 'desc'}


In [20]:
null_pct = accepted_ca_df_reduced.isna().mean() * 100

cols_to_drop = null_pct[null_pct > 50].index

print("Columns to drop (>50% missing):", len(cols_to_drop))

accepted_ca_df_reduced.drop(columns=cols_to_drop, inplace=True)

print("Shape after null drop:", accepted_ca_df_reduced.shape)


Columns to drop (>50% missing): 0
Shape after null drop: (314533, 101)


In [24]:
constant_cols = accepted_ca_df_reduced.columns[
    accepted_ca_df_reduced.nunique() <= 1
]

accepted_ca_df_reduced.drop(columns=constant_cols, inplace=True)
print("Shape after constant-column drop:", accepted_ca_df_reduced.shape)

accepted_ca_df_reduced.info(memory_usage='deep')


Shape after constant-column drop: (314533, 99)
<class 'pandas.core.frame.DataFrame'>
Index: 314533 entries, 13 to 2260696
Data columns (total 99 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   loan_amnt                   314533 non-null  float64
 1   funded_amnt                 314533 non-null  float64
 2   funded_amnt_inv             314533 non-null  float64
 3   term                        314533 non-null  object 
 4   int_rate                    314533 non-null  float64
 5   installment                 314533 non-null  float64
 6   grade                       314533 non-null  object 
 7   sub_grade                   314533 non-null  object 
 8   emp_length                  296393 non-null  object 
 9   home_ownership              314533 non-null  object 
 10  annual_inc                  314533 non-null  float64
 11  verification_status         314533 non-null  object 
 12  issue_d                     

In [25]:
null_pct = accepted_ca_df_reduced.isna().mean() * 100
null_pct.sort_values(ascending=False).head(15)


,0
mths_since_rcnt_il,42.573911
all_util,40.262230
open_acc_6m,40.253646
open_il_24m,40.253646
open_il_12m,40.253646
open_rv_12m,40.253646
inq_last_12m,40.253646
open_rv_24m,40.253646
max_bal_bc,40.253646
inq_fi,40.253646


In [26]:
cols_to_drop = null_pct[null_pct > 50].index
print("Dropping", len(cols_to_drop), "columns")

accepted_ca_df_reduced.drop(columns=cols_to_drop, inplace=True)

print("Shape after >50% null drop:", accepted_ca_df_reduced.shape)


Dropping 0 columns
Shape after >50% null drop: (314533, 99)


In [27]:
for col in accepted_ca_df_reduced.select_dtypes(include='object'):
    accepted_ca_df_reduced[col] = accepted_ca_df_reduced[col].astype('category')


In [28]:
for col in accepted_ca_df_reduced.select_dtypes(include='float64'):
    accepted_ca_df_reduced[col] = pd.to_numeric(accepted_ca_df_reduced[col], downcast='float')


In [29]:
accepted_ca_df_reduced.info(memory_usage='deep')


<class 'pandas.core.frame.DataFrame'>
Index: 314533 entries, 13 to 2260696
Data columns (total 99 columns):
 #   Column                      Non-Null Count   Dtype   
---  ------                      --------------   -----   
 0   loan_amnt                   314533 non-null  float32 
 1   funded_amnt                 314533 non-null  float32 
 2   funded_amnt_inv             314533 non-null  float64 
 3   term                        314533 non-null  category
 4   int_rate                    314533 non-null  float32 
 5   installment                 314533 non-null  float32 
 6   grade                       314533 non-null  category
 7   sub_grade                   314533 non-null  category
 8   emp_length                  296393 non-null  category
 9   home_ownership              314533 non-null  category
 10  annual_inc                  314533 non-null  float64 
 11  verification_status         314533 non-null  category
 12  issue_d                     314533 non-null  category
 13  lo

In [30]:
print("Final shape:", accepted_ca_df_reduced.shape)
print("Number of columns:", accepted_ca_df_reduced.shape[1])


Final shape: (314533, 99)
Number of columns: 99


In [31]:
accepted_ca_df_reduced.info(memory_usage='deep')


<class 'pandas.core.frame.DataFrame'>
Index: 314533 entries, 13 to 2260696
Data columns (total 99 columns):
 #   Column                      Non-Null Count   Dtype   
---  ------                      --------------   -----   
 0   loan_amnt                   314533 non-null  float32 
 1   funded_amnt                 314533 non-null  float32 
 2   funded_amnt_inv             314533 non-null  float64 
 3   term                        314533 non-null  category
 4   int_rate                    314533 non-null  float32 
 5   installment                 314533 non-null  float32 
 6   grade                       314533 non-null  category
 7   sub_grade                   314533 non-null  category
 8   emp_length                  296393 non-null  category
 9   home_ownership              314533 non-null  category
 10  annual_inc                  314533 non-null  float64 
 11  verification_status         314533 non-null  category
 12  issue_d                     314533 non-null  category
 13  lo

In [32]:
null_pct = accepted_ca_df_reduced.isna().mean() * 100
cols_to_drop = null_pct[null_pct > 50].index

print("Columns with >50% missing:", len(cols_to_drop))
print(cols_to_drop.tolist())

accepted_ca_df_reduced.drop(columns=cols_to_drop, inplace=True)

print("Final shape after null drop:", accepted_ca_df_reduced.shape)
accepted_ca_df_reduced.info(memory_usage='deep')


Columns with >50% missing: 0
[]
Final shape after null drop: (314533, 99)
<class 'pandas.core.frame.DataFrame'>
Index: 314533 entries, 13 to 2260696
Data columns (total 99 columns):
 #   Column                      Non-Null Count   Dtype   
---  ------                      --------------   -----   
 0   loan_amnt                   314533 non-null  float32 
 1   funded_amnt                 314533 non-null  float32 
 2   funded_amnt_inv             314533 non-null  float64 
 3   term                        314533 non-null  category
 4   int_rate                    314533 non-null  float32 
 5   installment                 314533 non-null  float32 
 6   grade                       314533 non-null  category
 7   sub_grade                   314533 non-null  category
 8   emp_length                  296393 non-null  category
 9   home_ownership              314533 non-null  category
 10  annual_inc                  314533 non-null  float64 
 11  verification_status         314533 non-null  c

In [33]:
leakage_cols = [
    'out_prncp','out_prncp_inv','total_pymnt','total_pymnt_inv',
    'total_rec_prncp','total_rec_int','total_rec_late_fee',
    'recoveries','collection_recovery_fee','last_pymnt_amnt',
    'last_pymnt_d','last_credit_pull_d',
    'last_fico_range_high','last_fico_range_low'
]

accepted_ca_df_reduced.drop(columns=[c for c in leakage_cols if c in accepted_ca_df_reduced], inplace=True)


In [35]:
import numpy as np

corr_matrix = accepted_ca_df_reduced.corr(numeric_only=True)
print(corr_matrix.shape)


(69, 69)


In [36]:
corr_abs = corr_matrix.abs()

# Keep only upper triangle (avoid duplicate pairs)
upper = corr_abs.where(np.triu(np.ones(corr_abs.shape), k=1).astype(bool))

# Find correlation pairs > 0.95
high_corr = [
    (col, row[row > 0.95].index.tolist(), row[row > 0.95].values.tolist())
    for col, row in upper.iterrows()
    if any(row > 0.95)
]

len(high_corr)


7

In [ ]:
accepted_df.columns


Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       ...
       'hardship_payoff_balance_amount', 'hardship_last_payment_amount',
       'disbursement_method', 'debt_settlement_flag',
       'debt_settlement_flag_date', 'settlement_status', 'settlement_date',
       'settlement_amount', 'settlement_percentage', 'settlement_term'],
      dtype='object', length=151)

In [ ]:
list(accepted_df.columns)


['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'annual_inc_joint',
 '

In [3]:
# 1. Make sure the original data is loaded (this might take a moment)
# accepted_df = pd.read_csv(accepted_path, low_memory=False)

# 2. Filter for California (Defining the missing variable)
accepted_ca_df = accepted_df[accepted_df['addr_state'] == 'CA'].copy()

# 3. NOW run the null value check
null_percentages = accepted_ca_df.isnull().sum() / len(accepted_ca_df)

# 4. Identify and drop columns with > 50% nulls
cols_to_drop = null_percentages[null_percentages > 0.50].index.tolist()
accepted_ca_df_cleaned = accepted_ca_df.drop(columns=cols_to_drop)

print(f"Original Columns: {len(accepted_ca_df.columns)}")
print(f"Columns Dropped: {len(cols_to_drop)}")
print(f"Remaining Columns: {len(accepted_ca_df_cleaned.columns)}")

NameError: name 'accepted_df' is not defined

# 3. What your NEXT step should be (inside Notebook 01)

You are now at the most important part of the whole project:

👉 Defining the modeling frame.

Your goal in this notebook is NOT to clean everything, but to:

✔ formally define “default”
✔ classify columns
✔ flag leakage
✔ design your reduced variable list

### Step 3.1 — Inspect and lock the target

In [ ]:
accepted_df["loan_status"].value_counts(dropna=False)


,count
loan_status,
Fully Paid,1076751
Current,878317
Charged Off,268559
Late (31-120 days),21467
In Grace Period,8436
Late (16-30 days),4349
Does not meet the credit policy. Status:Fully Paid,1988
Does not meet the credit policy. Status:Charged Off,761
Default,40


# Step 2: Create a binary column target_default

In [ ]:
# First, define mapping
default_status = [
    "Charged Off",
    "Late (31-120 days)",
    "Does not meet the credit policy. Status:Charged Off",
    "Default"
]

non_default_status = [
    "Fully Paid",
    "Does not meet the credit policy. Status:Fully Paid"
]

# Apply mapping
def map_default(status):
    if status in default_status:
        return 1
    elif status in non_default_status:
        return 0
    else:
        return np.nan  # Current, Late 16-30, In Grace, NaN

accepted_df["target_default"] = accepted_df["loan_status"].apply(map_default)

# Check counts
accepted_df["target_default"].value_counts(dropna=False)


,count
target_default,
0.0,1078739
NaN,891135
1.0,290827


### modeling population

For modeling, We only want fully labeled loans:

Include: target_default = 0 or 1

Exclude: target_default = NaN (891k rows)

In [ ]:
accepted_model_df = accepted_df[accepted_df["target_default"].notna()].copy()
accepted_model_df.shape


(1369566, 152)

0.0 + 1.0 = 1,078,739 + 290,827 = 1,369,566 rows

Save this checked dataset

Even though we haven’t cleaned predictors yet, save this labeled set to processed folder:

In [ ]:
accepted_model_df.to_csv(PROCESSED_PATH + "accepted_checked.csv", index=False)


In [ ]:
column_class = {
    # ID / metadata
    "id": "ID",
    "member_id": "ID",
    "url": "ID",
    "desc": "ID",
    "title": "ID",
    "emp_title": "ID",  # Too free text to use directly

    # Target
    "loan_status": "Original Target",
    "target_default": "Target",

    # Leakage / post-loan
    "out_prncp": "Leakage",
    "out_prncp_inv": "Leakage",
    "total_pymnt": "Leakage",
    "total_pymnt_inv": "Leakage",
    "total_rec_prncp": "Leakage",
    "total_rec_int": "Leakage",
    "total_rec_late_fee": "Leakage",
    "recoveries": "Leakage",
    "collection_recovery_fee": "Leakage",
    "last_pymnt_d": "Leakage",
    "last_pymnt_amnt": "Leakage",
    "next_pymnt_d": "Leakage",
    "last_credit_pull_d": "Leakage",
    "last_fico_range_high": "Leakage",
    "last_fico_range_low": "Leakage",
    "hardship_flag": "Leakage",
    "hardship_type": "Leakage",
    "hardship_reason": "Leakage",
    "hardship_status": "Leakage",
    "deferral_term": "Leakage",
    "hardship_amount": "Leakage",
    "hardship_start_date": "Leakage",
    "hardship_end_date": "Leakage",
    "payment_plan_start_date": "Leakage",
    "hardship_length": "Leakage",
    "hardship_dpd": "Leakage",
    "hardship_loan_status": "Leakage",
    "orig_projected_additional_accrued_interest": "Leakage",
    "hardship_payoff_balance_amount": "Leakage",
    "hardship_last_payment_amount": "Leakage",
    "disbursement_method": "Leakage",
    "debt_settlement_flag": "Leakage",
    "debt_settlement_flag_date": "Leakage",
    "settlement_status": "Leakage",
    "settlement_date": "Leakage",
    "settlement_amount": "Leakage",
    "settlement_percentage": "Leakage",
    "settlement_term": "Leakage",

    # Application-time predictors
    "loan_amnt": "Predictor",
    "funded_amnt": "Predictor",
    "funded_amnt_inv": "Predictor",
    "term": "Predictor",
    "int_rate": "Predictor",
    "installment": "Predictor",
    "grade": "Predictor",
    "sub_grade": "Predictor",
    "emp_length": "Predictor",
    "home_ownership": "Predictor",
    "annual_inc": "Predictor",
    "verification_status": "Predictor",
    "issue_d": "Predictor",
    "pymnt_plan": "Predictor",
    "purpose": "Predictor",
    "zip_code": "Predictor",
    "addr_state": "Predictor",
    "dti": "Predictor",
    "delinq_2yrs": "Predictor",
    "earliest_cr_line": "Predictor",
    "fico_range_low": "Predictor",
    "fico_range_high": "Predictor",
    "inq_last_6mths": "Predictor",
    "mths_since_last_delinq": "Predictor",
    "mths_since_last_record": "Predictor",
    "open_acc": "Predictor",
    "pub_rec": "Predictor",
    "revol_bal": "Predictor",
    "revol_util": "Predictor",
    "total_acc": "Predictor",
    "initial_list_status": "Predictor",
    "collections_12_mths_ex_med": "Predictor",
    "mths_since_last_major_derog": "Predictor",
    "policy_code": "Predictor",
    "application_type": "Predictor",
    "annual_inc_joint": "Predictor",
    "dti_joint": "Predictor",
    "verification_status_joint": "Predictor",
    "acc_now_delinq": "Predictor",
    "tot_coll_amt": "Predictor",
    "tot_cur_bal": "Predictor",
    "open_acc_6m": "Predictor",
    "open_act_il": "Predictor",
    "open_il_12m": "Predictor",
    "open_il_24m": "Predictor",
    "mths_since_rcnt_il": "Predictor",
    "total_bal_il": "Predictor",
    "il_util": "Predictor",
    "open_rv_12m": "Predictor",
    "open_rv_24m": "Predictor",
    "max_bal_bc": "Predictor",
    "all_util": "Predictor",
    "total_rev_hi_lim": "Predictor",
    "inq_fi": "Predictor",
    "total_cu_tl": "Predictor",
    "inq_last_12m": "Predictor",
    "acc_open_past_24mths": "Predictor",
    "avg_cur_bal": "Predictor",
    "bc_open_to_buy": "Predictor",
    "bc_util": "Predictor",
    "chargeoff_within_12_mths": "Predictor",
    "delinq_amnt": "Predictor",
    "mo_sin_old_il_acct": "Predictor",
    "mo_sin_old_rev_tl_op": "Predictor",
    "mo_sin_rcnt_rev_tl_op": "Predictor",
    "mo_sin_rcnt_tl": "Predictor",
    "mort_acc": "Predictor",
    "mths_since_recent_bc": "Predictor",
    "mths_since_recent_bc_dlq": "Predictor",
    "mths_since_recent_inq": "Predictor",
    "mths_since_recent_revol_delinq": "Predictor",
    "num_accts_ever_120_pd": "Predictor",
    "num_actv_bc_tl": "Predictor",
    "num_actv_rev_tl": "Predictor",
    "num_bc_sats": "Predictor",
    "num_bc_tl": "Predictor",
    "num_il_tl": "Predictor",
    "num_op_rev_tl": "Predictor",
    "num_rev_accts": "Predictor",
    "num_rev_tl_bal_gt_0": "Predictor",
    "num_sats": "Predictor",
    "num_tl_120dpd_2m": "Predictor",
    "num_tl_30dpd": "Predictor",
    "num_tl_90g_dpd_24m": "Predictor",
    "num_tl_op_past_12m": "Predictor",
    "pct_tl_nvr_dlq": "Predictor",
    "percent_bc_gt_75": "Predictor",
    "pub_rec_bankruptcies": "Predictor",
    "tax_liens": "Predictor",
    "tot_hi_cred_lim": "Predictor",
    "total_bal_ex_mort": "Predictor",
    "total_bc_limit": "Predictor",
    "total_il_high_credit_limit": "Predictor",
    "revol_bal_joint": "Predictor",
    "sec_app_fico_range_low": "Predictor",
    "sec_app_fico_range_high": "Predictor",
    "sec_app_earliest_cr_line": "Predictor",
    "sec_app_inq_last_6mths": "Predictor",
    "sec_app_mort_acc": "Predictor",
    "sec_app_open_acc": "Predictor",
    "sec_app_revol_util": "Predictor",
    "sec_app_open_act_il": "Predictor",
    "sec_app_num_rev_accts": "Predictor",
    "sec_app_chargeoff_within_12_mths": "Predictor",
    "sec_app_collections_12_mths_ex_med": "Predictor",
    "sec_app_mths_since_last_major_derog": "Predictor",
    "hardship_flag": "Leakage",
    # ... remaining hardship and settlement columns → Leakage
}



In [ ]:
print(column_class)

{'id': 'ID', 'member_id': 'ID', 'url': 'ID', 'desc': 'ID', 'title': 'ID', 'emp_title': 'ID', 'loan_status': 'Original Target', 'target_default': 'Target', 'out_prncp': 'Leakage', 'out_prncp_inv': 'Leakage', 'total_pymnt': 'Leakage', 'total_pymnt_inv': 'Leakage', 'total_rec_prncp': 'Leakage', 'total_rec_int': 'Leakage', 'total_rec_late_fee': 'Leakage', 'recoveries': 'Leakage', 'collection_recovery_fee': 'Leakage', 'last_pymnt_d': 'Leakage', 'last_pymnt_amnt': 'Leakage', 'next_pymnt_d': 'Leakage', 'last_credit_pull_d': 'Leakage', 'last_fico_range_high': 'Leakage', 'last_fico_range_low': 'Leakage', 'hardship_flag': 'Leakage', 'hardship_type': 'Leakage', 'hardship_reason': 'Leakage', 'hardship_status': 'Leakage', 'deferral_term': 'Leakage', 'hardship_amount': 'Leakage', 'hardship_start_date': 'Leakage', 'hardship_end_date': 'Leakage', 'payment_plan_start_date': 'Leakage', 'hardship_length': 'Leakage', 'hardship_dpd': 'Leakage', 'hardship_loan_status': 'Leakage', 'orig_projected_additional_

In [ ]:
import pandas as pd

column_class_df = pd.DataFrame(list(column_class.items()), columns=["column","category"])
column_class_df.to_csv(PROCESSED_PATH + "column_classification.csv", index=False)


In [ ]:
print(column_class_df)

                                         column         category
0                                            id               ID
1                                     member_id               ID
2                                           url               ID
3                                          desc               ID
4                                         title               ID
5                                     emp_title               ID
6                                   loan_status  Original Target
7                                target_default           Target
8                                     out_prncp          Leakage
9                                 out_prncp_inv          Leakage
10                                  total_pymnt          Leakage
11                              total_pymnt_inv          Leakage
12                              total_rec_prncp          Leakage
13                                total_rec_int          Leakage
14                       

In [ ]:
import pandas as pd

# Full column classification
column_class = {
    # ID / metadata
    "id": "ID",
    "member_id": "ID",
    "url": "ID",
    "desc": "ID",
    "title": "ID",
    "emp_title": "ID",  # free text, too messy to use

    # Target
    "loan_status": "Original Target",
    "target_default": "Target",

    # Leakage / post-loan information
    "out_prncp": "Leakage",
    "out_prncp_inv": "Leakage",
    "total_pymnt": "Leakage",
    "total_pymnt_inv": "Leakage",
    "total_rec_prncp": "Leakage",
    "total_rec_int": "Leakage",
    "total_rec_late_fee": "Leakage",
    "recoveries": "Leakage",
    "collection_recovery_fee": "Leakage",
    "last_pymnt_d": "Leakage",
    "last_pymnt_amnt": "Leakage",
    "next_pymnt_d": "Leakage",
    "last_credit_pull_d": "Leakage",
    "last_fico_range_high": "Leakage",
    "last_fico_range_low": "Leakage",
    "hardship_flag": "Leakage",
    "hardship_type": "Leakage",
    "hardship_reason": "Leakage",
    "hardship_status": "Leakage",
    "deferral_term": "Leakage",
    "hardship_amount": "Leakage",
    "hardship_start_date": "Leakage",
    "hardship_end_date": "Leakage",
    "payment_plan_start_date": "Leakage",
    "hardship_length": "Leakage",
    "hardship_dpd": "Leakage",
    "hardship_loan_status": "Leakage",
    "orig_projected_additional_accrued_interest": "Leakage",
    "hardship_payoff_balance_amount": "Leakage",
    "hardship_last_payment_amount": "Leakage",
    "disbursement_method": "Leakage",
    "debt_settlement_flag": "Leakage",
    "debt_settlement_flag_date": "Leakage",
    "settlement_status": "Leakage",
    "settlement_date": "Leakage",
    "settlement_amount": "Leakage",
    "settlement_percentage": "Leakage",
    "settlement_term": "Leakage",

    # Application-time predictors
    "loan_amnt": "Predictor",
    "funded_amnt": "Predictor",
    "funded_amnt_inv": "Predictor",
    "term": "Predictor",
    "int_rate": "Predictor",
    "installment": "Predictor",
    "grade": "Predictor",
    "sub_grade": "Predictor",
    "emp_length": "Predictor",
    "home_ownership": "Predictor",
    "annual_inc": "Predictor",
    "verification_status": "Predictor",
    "issue_d": "Predictor",
    "pymnt_plan": "Predictor",
    "purpose": "Predictor",
    "zip_code": "Predictor",
    "addr_state": "Predictor",
    "dti": "Predictor",
    "delinq_2yrs": "Predictor",
    "earliest_cr_line": "Predictor",
    "fico_range_low": "Predictor",
    "fico_range_high": "Predictor",
    "inq_last_6mths": "Predictor",
    "mths_since_last_delinq": "Predictor",
    "mths_since_last_record": "Predictor",
    "open_acc": "Predictor",
    "pub_rec": "Predictor",
    "revol_bal": "Predictor",
    "revol_util": "Predictor",
    "total_acc": "Predictor",
    "initial_list_status": "Predictor",
    "collections_12_mths_ex_med": "Predictor",
    "mths_since_last_major_derog": "Predictor",
    "policy_code": "Predictor",
    "application_type": "Predictor",
    "annual_inc_joint": "Predictor",
    "dti_joint": "Predictor",
    "verification_status_joint": "Predictor",
    "acc_now_delinq": "Predictor",
    "tot_coll_amt": "Predictor",
    "tot_cur_bal": "Predictor",
    "open_acc_6m": "Predictor",
    "open_act_il": "Predictor",
    "open_il_12m": "Predictor",
    "open_il_24m": "Predictor",
    "mths_since_rcnt_il": "Predictor",
    "total_bal_il": "Predictor",
    "il_util": "Predictor",
    "open_rv_12m": "Predictor",
    "open_rv_24m": "Predictor",
    "max_bal_bc": "Predictor",
    "all_util": "Predictor",
    "total_rev_hi_lim": "Predictor",
    "inq_fi": "Predictor",
    "total_cu_tl": "Predictor",
    "inq_last_12m": "Predictor",
    "acc_open_past_24mths": "Predictor",
    "avg_cur_bal": "Predictor",
    "bc_open_to_buy": "Predictor",
    "bc_util": "Predictor",
    "chargeoff_within_12_mths": "Predictor",
    "delinq_amnt": "Predictor",
    "mo_sin_old_il_acct": "Predictor",
    "mo_sin_old_rev_tl_op": "Predictor",
    "mo_sin_rcnt_rev_tl_op": "Predictor",
    "mo_sin_rcnt_tl": "Predictor",
    "mort_acc": "Predictor",
    "mths_since_recent_bc": "Predictor",
    "mths_since_recent_bc_dlq": "Predictor",
    "mths_since_recent_inq": "Predictor",
    "mths_since_recent_revol_delinq": "Predictor",
    "num_accts_ever_120_pd": "Predictor",
    "num_actv_bc_tl": "Predictor",
    "num_actv_rev_tl": "Predictor",
    "num_bc_sats": "Predictor",
    "num_bc_tl": "Predictor",
    "num_il_tl": "Predictor",
    "num_op_rev_tl": "Predictor",
    "num_rev_accts": "Predictor",
    "num_rev_tl_bal_gt_0": "Predictor",
    "num_sats": "Predictor",
    "num_tl_120dpd_2m": "Predictor",
    "num_tl_30dpd": "Predictor",
    "num_tl_90g_dpd_24m": "Predictor",
    "num_tl_op_past_12m": "Predictor",
    "pct_tl_nvr_dlq": "Predictor",
    "percent_bc_gt_75": "Predictor",
    "pub_rec_bankruptcies": "Predictor",
    "tax_liens": "Predictor",
    "tot_hi_cred_lim": "Predictor",
    "total_bal_ex_mort": "Predictor",
    "total_bc_limit": "Predictor",
    "total_il_high_credit_limit": "Predictor",
    "revol_bal_joint": "Predictor",
    "sec_app_fico_range_low": "Predictor",
    "sec_app_fico_range_high": "Predictor",
    "sec_app_earliest_cr_line": "Predictor",
    "sec_app_inq_last_6mths": "Predictor",
    "sec_app_mort_acc": "Predictor",
    "sec_app_open_acc": "Predictor",
    "sec_app_revol_util": "Predictor",
    "sec_app_open_act_il": "Predictor",
    "sec_app_num_rev_accts": "Predictor",
    "sec_app_chargeoff_within_12_mths": "Predictor",
    "sec_app_collections_12_mths_ex_med": "Predictor",
    "sec_app_mths_since_last_major_derog": "Predictor"
}

# Convert to dataframe and save
column_class_df = pd.DataFrame(list(column_class.items()), columns=["column","category"])

# Update this path to your Google Drive processed folder
PROCESSED_PATH = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/processed/"
column_class_df.to_csv(PROCESSED_PATH + "column_classification.csv", index=False)

print("Column classification dictionary saved successfully!")


Column classification dictionary saved successfully!


### Extract Calfornia Data from the rejected_2007_to_2018Q4.csv

In [ ]:
# Import necessary libraries
import pandas as pd
from google.colab import drive

# Mount Google Drive if not already mounted
drive.mount('/content/drive')

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
RAW_PATH = PROJECT_ROOT + "data/raw/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Load raw rejected dataset
rejected_raw_file = RAW_PATH + "rejected_2007_to_2018Q4.csv"
rejected_df = pd.read_csv(rejected_raw_file, low_memory=False)

# Check columns
print(rejected_df.columns)

# Filter California loans
rejected_ca_df = rejected_df[rejected_df["addr_state"] == "CA"].copy()

# Check shape
print("California-only rejected loans shape:", rejected_ca_df.shape)

# Save
rejected_ca_df.to_csv(PROCESSED_PATH + "rejected_CA.csv", index=False)



Index(['Amount Requested', 'Application Date', 'Loan Title', 'Risk_Score',
       'Debt-To-Income Ratio', 'Zip Code', 'State', 'Employment Length',
       'Policy Code'],
      dtype='object')


KeyError: 'addr_state'

In [ ]:
# Filter California loans using 'State' column
rejected_ca_df = rejected_df[rejected_df["State"] == "CA"].copy()

# Check shape
print("California-only rejected loans shape:", rejected_ca_df.shape)

# Save
rejected_ca_df.to_csv(PROCESSED_PATH + "rejected_CA.csv", index=False)


California-only rejected loans shape: (3242169, 9)


In [ ]:
# Add a column to indicate accepted vs rejected
ca_df['loan_status'] = 'Accepted'
rejected_ca_df['loan_status'] = 'Rejected'

# Concatenate datasets
full_ca_loans = pd.concat([ca_df, rejected_ca_df], ignore_index=True)

# Check the combined shape
print("Full California loans dataset shape:", full_ca_loans.shape)
full_ca_loans.head()


NameError: name 'ca_df' is not defined

### Marge both datasets

In [ ]:
import pandas as pd
from google.colab import drive

# Mount Google Drive
drive.mount('/content/drive')

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"

# Load California accepted loans
ca_df = pd.read_csv(PROCESSED_PATH + "accepted_CA.csv")

# Load California rejected loans
rejected_ca_df = pd.read_csv(PROCESSED_PATH + "rejected_CA.csv")

# Add loan status
ca_df['loan_status'] = 'Accepted'
rejected_ca_df['loan_status'] = 'Rejected'

# Concatenate datasets
full_ca_loans = pd.concat([ca_df, rejected_ca_df], ignore_index=True)

# Check shape
print("Full California loans dataset shape:", full_ca_loans.shape)
full_ca_loans.head()


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/tmp/ipython-input-49850068.py:12: DtypeWarning: Columns (19,49,59,118,129,130,131,134,135,136,139) have mixed types. Specify dtype option on import or set low_memory=False.
  ca_df = pd.read_csv(PROCESSED_PATH + "accepted_CA.csv")


Full California loans dataset shape: (3442021, 161)


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,target_default,Amount Requested,Application Date,Loan Title,Risk_Score,Debt-To-Income Ratio,Zip Code,State,Employment Length,Policy Code
0,66624733.0,NaN,18000.0,18000.0,18000.0,60 months,19.48,471.70,E,E2,...,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,68466066.0,NaN,20200.0,20200.0,20200.0,60 months,18.49,518.35,D,D5,...,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,68606528.0,NaN,12000.0,12000.0,12000.0,36 months,9.17,382.55,B,B2,...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,68356797.0,NaN,15975.0,15975.0,15975.0,36 months,16.59,566.30,D,D2,...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,68406292.0,NaN,22875.0,22875.0,22875.0,60 months,15.77,553.49,D,D1,...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


save the marge dataset in filtered folder

In [ ]:
import os

# Create filtered folder if it doesn't exist
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"
os.makedirs(FILTERED_PATH, exist_ok=True)

# Save full California dataset
full_ca_loans.to_csv(FILTERED_PATH + "full_CA_dataset.csv", index=False)

print("Full California dataset saved successfully at:", FILTERED_PATH + "full_CA_dataset.csv")


Full California dataset saved successfully at: /content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/full_CA_dataset.csv


### Marge

In [ ]:
import pandas as pd
from google.colab import drive

# Mount Google Drive (if not already mounted)
drive.mount('/content/drive', force_remount=True)

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"

# Load full CA loans dataset
full_ca_df = pd.read_csv(FILTERED_PATH + "full_CA_dataset.csv", low_memory=False)

# Load 3rd dataset (e.g., CA unemployment)
unemployment_ca_df = pd.read_csv(PROCESSED_PATH + "ca_unemployment.csv", low_memory=False)

# Check the columns to decide merge keys
print("Full CA loans columns:", full_ca_df.columns)
print("Unemployment CA columns:", unemployment_ca_df.columns)


Mounted at /content/drive


In [ ]:
import pandas as pd
from google.colab import drive

# Mount Google Drive (force remount if needed)
drive.mount('/content/drive', force_remount=True)

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"

# Load full California loans dataset (accepted + rejected)
full_ca_df = pd.read_csv(FILTERED_PATH + "full_CA_dataset.csv", low_memory=False)

# Load 3rd dataset (e.g., CA unemployment)
unemployment_ca_df = pd.read_csv(PROCESSED_PATH + "ca_unemployment.csv", low_memory=False)

# Check columns to decide merge keys
print("Full CA loans columns:", full_ca_df.columns)
print("Unemployment CA columns:", unemployment_ca_df.columns)

# Merge unemployment data with full CA loans
# Replace 'year' and 'month' with actual column names if different
full_ca_merged = full_ca_df.merge(unemployment_ca_df, on=['year', 'month'], how='left')

# Check the shape and preview
print("Merged dataset shape:", full_ca_merged.shape)
full_ca_merged.head()

# Save the merged dataset
full_ca_merged.to_csv(FILTERED_PATH + "full_CA_merged_dataset.csv", index=False)
print("Merged dataset saved successfully at:", FILTERED_PATH + "full_CA_merged_dataset.csv")



Mounted at /content/drive


In [ ]:
import pandas as pd
from google.colab import drive
import os

# Mount Google Drive
drive.mount('/content/drive', force_remount=True)

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
RAW_PATH = PROJECT_ROOT + "data/raw/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"

# Make sure filtered folder exists
os.makedirs(FILTERED_PATH, exist_ok=True)


Mounted at /content/drive


In [ ]:
# Reload California accepted loans
ca_df = pd.read_csv(PROCESSED_PATH + "accepted_CA.csv", low_memory=False)

# Reload California rejected loans
rejected_ca_df = pd.read_csv(PROCESSED_PATH + "rejected_CA.csv", low_memory=False)

# Reload full CA dataset (accepted + rejected) if already saved
# full_ca_df = pd.read_csv(FILTERED_PATH + "full_CA_dataset.csv", low_memory=False)

# Reload 3rd dataset (e.g., CA unemployment)
unemployment_ca_df = pd.read_csv(PROCESSED_PATH + "ca_unemployment.csv", low_memory=False)


FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/processed/ca_unemployment.csv'

In [ ]:
import os

# List all files in processed folder
print(os.listdir(PROCESSED_PATH))


['accepted_CA.csv', 'accepted_checked.csv', 'column_classification_partial.csv', 'column_classification.csv', 'rejected_monthly_clean.csv', 'ca_unemployment_clean.csv', 'accepted_clean_model.csv', 'accepted_clean_model.parquet', 'accepted_sample_50k.parquet', 'reduced_accepted.csv', 'reduced_accepted.parquet', 'accepted_CA_clean_model.parquet', 'rejected_CA.csv']


In [ ]:
import pandas as pd

# Load CA unemployment data (correct file name)
unemployment_ca_df = pd.read_csv(PROCESSED_PATH + "ca_unemployment_clean.csv", low_memory=False)

# Preview columns to check merge keys
print("Unemployment CA columns:", unemployment_ca_df.columns)

# Make sure full_CA_dataset has 'year' and 'month' columns
full_ca_df['Application Date'] = pd.to_datetime(full_ca_df['Application Date'])
full_ca_df['year'] = full_ca_df['Application Date'].dt.year
full_ca_df['month'] = full_ca_df['Application Date'].dt.month

# Merge with unemployment
full_ca_merged = full_ca_df.merge(unemployment_ca_df, on=['year','month'], how='left')

# Save merged dataset
full_ca_merged.to_csv(FILTERED_PATH + "full_CA_merged_dataset.csv", index=False)
print("Merged dataset saved at:", FILTERED_PATH + "full_CA_merged_dataset.csv")


Unemployment CA columns: Index(['year_month', 'unemployment_rate'], dtype='object')


NameError: name 'full_ca_df' is not defined

In [ ]:
import pandas as pd
from google.colab import drive
import os

# Mount Drive
drive.mount('/content/drive', force_remount=True)

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"

# Ensure filtered folder exists
os.makedirs(FILTERED_PATH, exist_ok=True)


Mounted at /content/drive


In [ ]:
# Load accepted CA loans
ca_df = pd.read_csv(PROCESSED_PATH + "accepted_CA.csv", low_memory=False)

# Load rejected CA loans
rejected_ca_df = pd.read_csv(PROCESSED_PATH + "rejected_CA.csv", low_memory=False)

# Add loan status
ca_df['loan_status'] = 'Accepted'
rejected_ca_df['loan_status'] = 'Rejected'

# Concatenate to form full CA dataset
full_ca_df = pd.concat([ca_df, rejected_ca_df], ignore_index=True)

# Optional: check shape
print("Full CA dataset shape:", full_ca_df.shape)


Full CA dataset shape: (3442021, 161)


In [ ]:
# Load CA unemployment data
unemployment_ca_df = pd.read_csv(PROCESSED_PATH + "ca_unemployment_clean.csv", low_memory=False)

# Split 'year_month' into 'year' and 'month'
unemployment_ca_df['year'] = unemployment_ca_df['year_month'].str[:4].astype(int)
unemployment_ca_df['month'] = unemployment_ca_df['year_month'].str[5:7].astype(int)

# Keep only needed columns
unemployment_ca_df = unemployment_ca_df[['year', 'month', 'unemployment_rate']]


In [ ]:
# Convert Application Date to datetime
full_ca_df['Application Date'] = pd.to_datetime(full_ca_df['Application Date'])

# Extract year and month
full_ca_df['year'] = full_ca_df['Application Date'].dt.year
full_ca_df['month'] = full_ca_df['Application Date'].dt.month


In [ ]:
# Merge full CA loans with unemployment
full_ca_merged = full_ca_df.merge(unemployment_ca_df, on=['year','month'], how='left')

# Check shape and preview
print("Merged dataset shape:", full_ca_merged.shape)
full_ca_merged.head()


In [ ]:
merged_chunks = []

for y in unemployment_ca_df['year'].unique():
    for m in unemployment_ca_df['month'].unique():
        # Filter loans for this year & month
        temp_loans = full_ca_df[(full_ca_df['year'] == y) & (full_ca_df['month'] == m)]
        # Filter unemployment for this year & month
        temp_unemp = unemployment_ca_df[(unemployment_ca_df['year'] == y) & (unemployment_ca_df['month'] == m)]
        # Merge small chunk
        temp_merged = temp_loans.merge(temp_unemp, on=['year', 'month'], how='left')
        merged_chunks.append(temp_merged)

# Concatenate all chunks
full_ca_merged = pd.concat(merged_chunks, ignore_index=True)
print("Merged dataset shape:", full_ca_merged.shape)


NameError: name 'unemployment_ca_df' is not defined

In [ ]:
import pandas as pd
from google.colab import drive

# Mount Google Drive
drive.mount('/content/drive', force_remount=True)

# Paths
FILTERED_PATH = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/"

# Load full CA dataset
full_ca_df = pd.read_csv(FILTERED_PATH + "full_CA_dataset.csv", low_memory=False)

# Load CA unemployment dataset
unemployment_ca_df = pd.read_csv(FILTERED_PATH + "ca_unemployment_clean.csv", low_memory=False)

# Inspect columns
print("Columns in full_CA_dataset.csv:")
print(full_ca_df.columns)

print("\nColumns in ca_unemployment_clean.csv:")
print(unemployment_ca_df.columns)

# Preview first few rows
print("\nFull CA dataset preview:")
print(full_ca_df[['Application Date']].head())

print("\nCA Unemployment dataset preview:")
print(unemployment_ca_df.head())


Mounted at /content/drive


In [ ]:
import pandas as pd
from google.colab import drive

# Mount Google Drive
drive.mount('/content/drive', force_remount=True)

# Paths
FILTERED_PATH = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/"

# Load just the first few rows to inspect columns
full_ca_df_sample = pd.read_csv(FILTERED_PATH + "full_CA_dataset.csv", nrows=5)
unemployment_ca_df_sample = pd.read_csv(FILTERED_PATH + "ca_unemployment_clean.csv", nrows=5)

# Inspect columns
print("Columns in full_CA_dataset.csv:")
print(full_ca_df_sample.columns)

print("\nColumns in ca_unemployment_clean.csv:")
print(unemployment_ca_df_sample.columns)

# Preview first rows
print("\nFull CA dataset sample:")
print(full_ca_df_sample.head())

print("\nCA Unemployment dataset sample:")
print(unemployment_ca_df_sample.head())


Mounted at /content/drive
Columns in full_CA_dataset.csv:
Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       ...
       'target_default', 'Amount Requested', 'Application Date', 'Loan Title',
       'Risk_Score', 'Debt-To-Income Ratio', 'Zip Code', 'State',
       'Employment Length', 'Policy Code'],
      dtype='object', length=161)

Columns in ca_unemployment_clean.csv:
Index(['year_month', 'unemployment_rate'], dtype='object')

Full CA dataset sample:
           id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  66624733.0        NaN    18000.0      18000.0          18000.0   60 months   
1  68466066.0        NaN    20200.0      20200.0          20200.0   60 months   
2  68606528.0        NaN    12000.0      12000.0          12000.0   36 months   
3  68356797.0        NaN    15975.0      15975.0          15975.0   36 months   
4  68406292.0        NaN    22875.0      22

In [ ]:
# Ensure Application Date is datetime
full_ca_df['Application Date'] = pd.to_datetime(full_ca_df['Application Date'], errors='coerce')

# Drop rows without Application Date to reduce memory
full_ca_df = full_ca_df.dropna(subset=['Application Date'])

# Extract year and month from Application Date
full_ca_df['year'] = full_ca_df['Application Date'].dt.year
full_ca_df['month'] = full_ca_df['Application Date'].dt.month

# Extract year and month from unemployment dataset
unemployment_ca_df['year'] = unemployment_ca_df['year_month'].str[:4].astype(int)
unemployment_ca_df['month'] = unemployment_ca_df['year_month'].str[5:7].astype(int)

# Keep only necessary columns in unemployment
unemployment_ca_df = unemployment_ca_df[['year', 'month', 'unemployment_rate']]

# Check first few rows
print(full_ca_df[['Application Date','year','month']].head())
print(unemployment_ca_df.head())


NameError: name 'full_ca_df' is not defined

In [ ]:
import pandas as pd
from google.colab import drive
import os

# Mount Drive
drive.mount('/content/drive', force_remount=True)

# Paths
PROJECT_ROOT = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/"
FILTERED_PATH = PROJECT_ROOT + "data/filtered/"
PROCESSED_PATH = PROJECT_ROOT + "data/processed/"

# Ensure filtered folder exists
os.makedirs(FILTERED_PATH, exist_ok=True)

# Step 1: Load accepted + rejected CA loans
ca_df = pd.read_csv(FILTERED_PATH + "accepted_CA.csv", low_memory=False)
rejected_ca_df = pd.read_csv(FILTERED_PATH + "rejected_CA.csv", low_memory=False)

# Add loan status
ca_df['loan_status'] = 'Accepted'
rejected_ca_df['loan_status'] = 'Rejected'

# Combine datasets
full_ca_df = pd.concat([ca_df, rejected_ca_df], ignore_index=True)
print("Full CA dataset shape:", full_ca_df.shape)

# Step 2: Prepare full CA dataset for merge
full_ca_df['Application Date'] = pd.to_datetime(full_ca_df['Application Date'], errors='coerce')
full_ca_df = full_ca_df.dropna(subset=['Application Date'])
full_ca_df['year'] = full_ca_df['Application Date'].dt.year
full_ca_df['month'] = full_ca_df['Application Date'].dt.month

# Step 3: Load CA unemployment and prepare keys
unemployment_ca_df = pd.read_csv(FILTERED_PATH + "ca_unemployment_clean.csv", low_memory=False)
unemployment_ca_df['year'] = unemployment_ca_df['year_month'].str[:4].astype(int)
unemployment_ca_df['month'] = unemployment_ca_df['year_month'].str[5:7].astype(int)
unemployment_ca_df = unemployment_ca_df[['year','month','unemployment_rate']]

# Step 4: Merge in chunks to avoid memory crash
merged_chunks = []

for y, m in unemployment_ca_df[['year','month']].drop_duplicates().values:
    temp_loans = full_ca_df[(full_ca_df['year']==y) & (full_ca_df['month']==m)]
    temp_unemp = unemployment_ca_df[(unemployment_ca_df['year']==y) & (unemployment_ca_df['month']==m)]
    temp_merged = temp_loans.merge(temp_unemp, on=['year','month'], how='left')
    merged_chunks.append(temp_merged)

# Concatenate all chunks
full_ca_merged = pd.concat(merged_chunks, ignore_index=True)
print("Merged dataset shape:", full_ca_merged.shape)

# Step 5: Save final merged dataset
full_ca_merged.to_csv(FILTERED_PATH + "full_CA_merged_dataset.csv", index=False)
print("Merged dataset saved at:", FILTERED_PATH + "full_CA_merged_dataset.csv")


Mounted at /content/drive


FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/accepted_CA.csv'

In [ ]:
FILTERED_PATH = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/"


In [ ]:
import pandas as pd
from google.colab import drive
import os

# Mount Drive
drive.mount('/content/drive', force_remount=True)

# Path
FILTERED_PATH = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/"
os.makedirs(FILTERED_PATH, exist_ok=True)

# Step 1: Load accepted + rejected CA loans
ca_df = pd.read_csv(FILTERED_PATH + "accepted_CA.csv", low_memory=False)
rejected_ca_df = pd.read_csv(FILTERED_PATH + "rejected_CA.csv", low_memory=False)

# Add loan status
ca_df['loan_status'] = 'Accepted'
rejected_ca_df['loan_status'] = 'Rejected'

# Combine datasets
full_ca_df = pd.concat([ca_df, rejected_ca_df], ignore_index=True)
print("Full CA dataset shape:", full_ca_df.shape)

# Step 2: Prepare full CA dataset for merge
full_ca_df['Application Date'] = pd.to_datetime(full_ca_df['Application Date'], errors='coerce')
full_ca_df = full_ca_df.dropna(subset=['Application Date'])
full_ca_df['year'] = full_ca_df['Application Date'].dt.year
full_ca_df['month'] = full_ca_df['Application Date'].dt.month

# Step 3: Load CA unemployment and prepare keys
unemployment_ca_df = pd.read_csv(FILTERED_PATH + "ca_unemployment_clean.csv", low_memory=False)
unemployment_ca_df['year'] = unemployment_ca_df['year_month'].str[:4].astype(int)
unemployment_ca_df['month'] = unemployment_ca_df['year_month'].str[5:7].astype(int)
unemployment_ca_df = unemployment_ca_df[['year','month','unemployment_rate']]

# Step 4: Merge in chunks to avoid memory crash
merged_chunks = []

for y, m in unemployment_ca_df[['year','month']].drop_duplicates().values:
    temp_loans = full_ca_df[(full_ca_df['year']==y) & (full_ca_df['month']==m)]
    temp_unemp = unemployment_ca_df[(unemployment_ca_df['year']==y) & (unemployment_ca_df['month']==m)]
    temp_merged = temp_loans.merge(temp_unemp, on=['year','month'], how='left')
    merged_chunks.append(temp_merged)

# Concatenate all chunks
full_ca_merged = pd.concat(merged_chunks, ignore_index=True)
print("Merged dataset shape:", full_ca_merged.shape)

# Step 5: Save final merged dataset
full_ca_merged.to_csv(FILTERED_PATH + "full_CA_merged_dataset.csv", index=False)
print("Merged dataset saved at:", FILTERED_PATH + "full_CA_merged_dataset.csv")


Mounted at /content/drive
Full CA dataset shape: (3442021, 161)


In [ ]:
import pandas as pd
from google.colab import drive

drive.mount('/content/drive', force_remount=True)
FILTERED_PATH = "/content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/"

# Load only first 50k rows from accepted and rejected CA loans
ca_df = pd.read_csv(FILTERED_PATH + "accepted_CA.csv", nrows=50000, low_memory=False)
rejected_ca_df = pd.read_csv(FILTERED_PATH + "rejected_CA.csv", nrows=50000, low_memory=False)

# Add loan status
ca_df['loan_status'] = 'Accepted'
rejected_ca_df['loan_status'] = 'Rejected'

# Combine sample datasets
sample_ca_df = pd.concat([ca_df, rejected_ca_df], ignore_index=True)
print("Sample CA dataset shape:", sample_ca_df.shape)


Mounted at /content/drive
Sample CA dataset shape: (100000, 161)


In [ ]:
# Convert Application Date to datetime
sample_ca_df['Application Date'] = pd.to_datetime(sample_ca_df['Application Date'], errors='coerce')
sample_ca_df = sample_ca_df.dropna(subset=['Application Date'])
sample_ca_df['year'] = sample_ca_df['Application Date'].dt.year
sample_ca_df['month'] = sample_ca_df['Application Date'].dt.month


In [ ]:
unemployment_ca_df = pd.read_csv(FILTERED_PATH + "ca_unemployment_clean.csv", low_memory=False)
unemployment_ca_df['year'] = unemployment_ca_df['year_month'].str[:4].astype(int)
unemployment_ca_df['month'] = unemployment_ca_df['year_month'].str[5:7].astype(int)
unemployment_ca_df = unemployment_ca_df[['year','month','unemployment_rate']]


In [ ]:
# Merge using a simple left join
sample_merged = sample_ca_df.merge(unemployment_ca_df, on=['year','month'], how='left')
print("Sample merged dataset shape:", sample_merged.shape)

# Preview
sample_merged.head()


Sample merged dataset shape: (50000, 164)


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,Loan Title,Risk_Score,Debt-To-Income Ratio,Zip Code,State,Employment Length,Policy Code,year,month,unemployment_rate
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,big fat lender,638.0,17.73%,941xx,CA,< 1 year,0.0,2007,5,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Good Heart Bad Credit!,463.0,12.34%,953xx,CA,1 year,0.0,2007,5,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Jlb7557,529.0,27.01%,925xx,CA,1 year,0.0,2007,6,NaN
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,Loan Consolidation,635.0,13.35%,926xx,CA,1 year,0.0,2007,6,NaN
4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,acordars,576.0,14.28%,926xx,CA,2 years,0.0,2007,7,NaN


In [ ]:
# Save the merged sample dataset
sample_merged.to_csv(FILTERED_PATH + "full_CA_merged_sample.csv", index=False)
print("Sample merged dataset saved at:", FILTERED_PATH + "full_CA_merged_sample.csv")


Sample merged dataset saved at: /content/drive/MyDrive/LOAN_DEFAULT_RISK/data/filtered/full_CA_merged_sample.csv
